### PR TITLE
Added more tests for game and ai

### PR DESCRIPTION
**Pull request by Vithop**

### CODE REVIEW

File: svelte/src/lib/Game.ts [hunk 1]
```diff
@@ -17,105 +17,103 @@ export const Status = {
 
 export type GameStatus = (typeof Status)[keyof typeof Status];
 
-function isBoardFull(boardState: string[]) {
-  return boardState.every((value) => value !== "");
-}
+export type TTTPiece = "X" | "O" | "";
```
Comment (Khufu-I) [inline svelte/src/lib/Game.ts:20]: In Kotlin we used the term `Player`. Ideally this is an optional enum and we don't need empty string.

File: svelte/src/lib/Game.ts [hunk 2]
```diff
@@ -17,105 +17,103 @@ export const Status = {
...
-  [0, 1, 2],
-  [3, 4, 5],
-  [6, 7, 8],
-  [0, 3, 6],
-  [1, 4, 7],
-  [2, 5, 8],
-  [0, 4, 8],
-  [2, 4, 6]
-];
+  private winMap = [
```
Comment (Khufu-I) [inline svelte/src/lib/Game.ts:25]: This could be a static variable outside of this class.

File: svelte/src/lib/Game.ts [hunk 3]
```diff
@@ -17,105 +17,103 @@ export const Status = {
...
-    return this.#boardState;
+  get board(): TTTPiece[] {
+    return this.#board;
   }
 
   get isGameOver(): boolean {
-    return (
-      this.#gameStatus !== Status.XTurn && this.#gameStatus != Status.OTurn
-    );
+    return this.#status >= Status.Draw;
```
Comment (Khufu-I) [inline svelte/src/lib/Game.ts:52]: This is dangerous, do we have tests for `isGameOver`

File: svelte/src/lib/Game.ts [hunk 4]
```diff
@@ -17,105 +17,103 @@ export const Status = {
...
-      : this.#boardState
-          .map((elem, index) =>
-            elem === availableSpot ? index : deleteElement
-          )
-          .filter((elem) => elem !== deleteElement);
+  get availableMoves(): number[] {
+    return this.#board.reduce<number[]>((acc, val, idx) => {
+      if (val === "") acc.push(idx);
+      return acc;
+    }, []);
```
Comment (Khufu-I) [inline svelte/src/lib/Game.ts:59]: I don't think this is correct. If the game is over before the board is full this returns the wrong results.

File: svelte/src/lib/Game.ts [hunk 5]
```diff
@@ -17,105 +17,103 @@ export const Status = {
...
     );
     console.assert(
-      this.#boardState[i] === "",
-      `Tile at index ${i} is already occupied with ${this.#boardState[i]}`
+      this.#board[i] === "",
+      `Tile at index ${i} is already occupied with ${this.#board[i]}`
     );
     // Log invalid moves and then ignore
-    if (this.isGameOver || this.#boardState[i] != "") return;
+    if (this.isGameOver || this.#board[i] != "") return this.#status;
```
Comment (Khufu-I) [inline svelte/src/lib/Game.ts:72]: Wonder if an exception should be thrown, that's what we're doing in Kotlin.

File: svelte/src/lib/Game.ts [hunk 6]
```diff
@@ -17,105 +17,103 @@ export const Status = {
...
+    if (this.#status == Status.XTurn) return Status.OTurn;
+    if (this.#status == Status.OTurn) return Status.XTurn;
+    return this.#status;
+  }
+
+  private get isBoardFull(): boolean {
+    return this.#board.every((value) => value !== "");
+  }
+
+  private isWinPattern(winPattern: number[]): boolean {
```
Comment (Khufu-I) [inline svelte/src/lib/Game.ts:112]: Why does this have to be member function. Free functions are often more desirable.

File: svelte/src/lib/Game.ts [hunk 7]
```diff
@@ -17,105 +17,103 @@ export const Status = {
...
-    if (this.#gameStatus == Status.XTurn) return Status.OTurn;
-    if (this.#gameStatus == Status.OTurn) return Status.XTurn;
-    return this.#gameStatus;
+    if (this.isBoardFull) return Status.Draw;
+    if (this.#status == Status.XTurn) return Status.OTurn;
+    if (this.#status == Status.OTurn) return Status.XTurn;
+    return this.#status;
+  }
+
+  private get isBoardFull(): boolean {
```
Comment (Khufu-I) [inline svelte/src/lib/Game.ts:108]: Why does this have to be member function. Free functions are often more desirable.

File: svelte/src/routes/TicTacToe/+page.svelte [hunk 8]
```diff
@@ -66,7 +65,7 @@
 
 <div class="game row flex-center">
   <div class="board {game.status == Status.XTurn ? 'x_turn' : 'o_turn'}">
-    {#each game.boardState as toe, i}
+    {#each game.board as toe, i}
```
Comment (Khufu-I) [inline svelte/src/routes/TicTacToe/+page.svelte:68]: 👍🏾

File: svelte/tests/lib/ai-player.spec.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+// Test AI Player
+
+import Game, { Status } from "../../src/lib/Game";
+import * as AIPlayer from "../../src/lib/AIPlayer";
+import { beforeEach, describe, expect, it } from "vitest";
+
+describe("AI Player", () => {
+  let tttGame: Game;
```
Comment (Khufu-I) [inline svelte/tests/lib/ai-player.spec.ts:8]: No need for a shared `Game` object just create it within each test block

File: svelte/tests/lib/ai-player.spec.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+  beforeEach(() => {
+    tttGame = new Game();
+  });
+
+  it("should select the best move on Hard", () => {
```
Comment (Khufu-I) [inline svelte/tests/lib/ai-player.spec.ts:14]: Good idea, we need similar tests for different difficulty levels on the Kotlin side.

File: svelte/tests/lib/game.spec.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
+import { beforeEach, describe, expect, it } from "vitest";
+import Game, { Status } from "../../src/lib/Game";
+
+describe("Tic Tac Toe Game", () => {
+  let ticTacToeGame: Game;
```
Comment (Khufu-I) [inline svelte/tests/lib/game.spec.ts:5]: It's better to localize this variable, it allows your tests to run in parallel.

File: svelte/tests/lib/game.spec.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
+import { beforeEach, describe, expect, it } from "vitest";
```
Comment (Khufu-I) [inline svelte/tests/lib/game.spec.ts:1]: I wonder if we should just use AI to generate tests from the AI code? If we use the same data types then this should be possible.

**Review by Khufu-I — CHANGES_REQUESTED** [review]

File: svelte/src/lib/Game.ts (hunk 1 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:20]: This is used for rendering purposes but I guess `"X" | "O" | undefined` would also work for both rendering and player

Comment (Vithop) [reply svelte/src/lib/Game.ts:20]: for typescript union type is better than enums

File: svelte/src/lib/Game.ts (hunk 2 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:25]: but i want it scoped into the game

File: svelte/src/lib/Game.ts (hunk 4 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:59]: tru then can we write this without branching?

File: svelte/src/lib/Game.ts (hunk 5 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:72]: Then we would need error handling. I'm thinking of making it more of a state machine which is more robust

File: svelte/src/lib/Game.ts (hunk 7 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:108]: i didn't know free funcs are desirable... explain?

File: svelte/src/lib/Game.ts (hunk 6 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:112]: same as above

File: svelte/tests/lib/game.spec.ts (hunk 12 above)
Comment (Vithop) [reply svelte/tests/lib/game.spec.ts:1]: let's do it

File: svelte/tests/lib/game.spec.ts (hunk 11 above)
Comment (Vithop) [reply svelte/tests/lib/game.spec.ts:5]: oh tru but it's a lot of duplicate code

File: svelte/src/lib/Game.ts (hunk 1 above)
Comment (Khufu-I) [reply svelte/src/lib/Game.ts:20]: The name should also be Player instead of `TTTPiece`

File: svelte/src/lib/Game.ts (hunk 7 above)
Comment (Khufu-I) [reply svelte/src/lib/Game.ts:108]: The main thinking is that free functions would use the public interface of objects vs internal data hence [leading to better encapsulation](https://www.drdobbs.com/cpp/how-non-member-functions-improve-encapsu/184401197).

File: svelte/src/lib/Game.ts (hunk 6 above)
Comment (Khufu-I) [reply svelte/src/lib/Game.ts:112]: Great [CppCon talk by Klaus Iglberger](https://www.youtube.com/watch?v=WLDT1lDOsb4) worth watching about this.

File: svelte/tests/lib/game.spec.ts (hunk 11 above)
Comment (Khufu-I) [reply svelte/tests/lib/game.spec.ts:5]: Tests are run many many times and creating a bit of duplicate code in order to enable parallelization is worth it.

**Review by Khufu-I — CHANGES_REQUESTED** [review]

File: svelte/src/lib/Game.ts (hunk 3 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:52]: But it's fast. There's three less operations

Comment (Vithop) [reply svelte/src/lib/Game.ts:52]: I also have tests cases for this but I could add more

File: svelte/src/lib/Game.ts (hunk 5 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:72]: I guess that's true will make an error for this

File: svelte/src/lib/Game.ts (hunk 7 above)
Comment (Vithop) [reply svelte/src/lib/Game.ts:108]: but this reads the board which is contained in the game?

File: svelte/src/lib/Game.ts [hunk 13]
```diff
@@ -53,23 +42,25 @@ export default class Game {
...
-      `Tile at index ${i} is already occupied with ${this.#board[i]}`
-    );
     // Log invalid moves and then ignore
-    if (this.isGameOver || this.#board[i] != "") return this.#status;
+    if (this.isGameOver)
+      throw new Error("Game is over. There are no available moves.");
+    if (this.#board[i] !== undefined)
+      throw new Error(
+        `Tile at index ${i} is already occupied with ${
+          this.#board[i] ?? "nothing"
```
Comment (Khufu-I) [inline svelte/src/lib/Game.ts:61]: Why would it be nothing? You're checking `this.#board[i] !== undefined`

File: svelte/tests/lib/game.spec.ts [hunk 14]
```diff
@@ -13,18 +13,48 @@ describe("Tic Tac Toe Game", () => {
...
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      undefined
+    ]);
```
Comment (Khufu-I) [inline svelte/tests/lib/game.spec.ts:26]: I believe an empty array `[]` is filled with undefined by default

**Review by Khufu-I — APPROVED** [review]

**Pull request merged**